src/controllers/media.ts

```diff
@@ -554,3 +337,131 @@ export const getByImdbID = async(ctx: ParameterizedContext<any, Router.IRouterPa
     throw new MediaNotFoundError();
   }
 };
+
+export const getVideo = async(ctx: ParameterizedContext<any, Router.IRouterParamContext<any, {}>>): Promise<any> => {
+  const { title, osdbHash, imdbID } = ctx.query;
+  let { episode, season, year, filebytesize } = ctx.query;
+  [episode, season, year, filebytesize] = [episode, season, year, filebytesize].map(param => param ? Number(param) : null);
+
+  if (!title && !osdbHash && !imdbID) {
+    throw new ValidationError('title, osdbHash or imdbId is a required parameter');
+  }
+
+  if (osdbHash && !filebytesize) {
+    throw new ValidationError('filebytesize is required when passing osdbHash');
+  }
+
+  const query = [];
+  const failedQuery = [];
+
+  if (osdbHash) {
+    query.push({ osdbHash });
+    failedQuery.push({ osdbHash });
+  }
+
+  if (imdbID) {
+    query.push({ imdbID });
+    failedQuery.push({ imdbId: imdbID });
+  }
+
+  if (title) {
+    const titleQuery: any = { searchMatches: { $in: [title] } };
+    const titleFailedQuery: any = { title };
+
+    if (year) {
+      titleQuery.year = year;
+      titleFailedQuery.year = year;
+    }
+    if (episode) {
+      titleQuery.episode = episode;
+      titleFailedQuery.episode = episode;
+    }
+    if (season) {
+      titleQuery.season = season;
+      titleFailedQuery.season = season;
+    }
+    query.push(titleQuery);
+    failedQuery.push(titleFailedQuery);
+  }
+
+  // find an existing metadata record, or previous failure record
+  const [existingResult, existingFailedResult] = await Promise.all([
+    MediaMetadata.findOne({ $or: query }, null, { lean: true }).exec(),
+    FailedLookups.findOne({ $or: failedQuery }, null, { lean: true }).exec(),
+  ]);
+
+  // we have an existing metadata record, so return it
+  if (existingResult) {
+    return ctx.body = existingResult;
+  }
+  // we have an existing failure record, so increment it, and throw not found error
+  if (existingFailedResult) {
+    await FailedLookups.updateOne({ _id: existingFailedResult._id }, { $inc: { count: 1 } }).exec();
+    throw new MediaNotFoundError();
+  }
+
+  // the database does not have a record of this file, so begin search for metadata on external apis.
+
+  // Start OpenSubtitles lookups
+  let openSubtitlesMetadata;
+
+  if (osdbHash && filebytesize) {
+    const osQuery = { moviehash: osdbHash, moviebytesize: filebytesize };
+    const validation = {
+      year: year ? year : null,
+      season: season ? season : null,
+      episode: episode ? episode : null,
+    };
+    openSubtitlesMetadata = await externalAPIHelper.getFromOpenSubtitles(osQuery, validation);
+    if (!openSubtitlesMetadata) {
+      await FailedLookups.updateOne({ osdbHash, imdbID, title, season, episode }, { $inc: { count: 1 } }, { upsert: true, setDefaultsOnInsert: true }).exec();
+      throw new MediaNotFoundError();
```

Comment:
Yep correct, so have made changes for that. I've left this here for the case where the client hasn't passed an imdbid, so it will still fail early, so added that extra check to line 416

Reply:
What if the client has sent a title and osdb hash?

Reply:
Yeah true. I guess this doesnt need to be here